plugins/inputs/sqlserver/azuresqlqueries.go

```diff
@@ -1217,3 +1217,275 @@ SELECT
 	,DATABASEPROPERTYEX(DB_NAME(), 'Updateability') as replica_updateability
 FROM sys.dm_os_schedulers AS s
 `
+
+const sqlAzureMIPartQueryPeriod = `
+DECLARE @QueryData NVARCHAR(MAX) = ?1; --input parameter
+
+IF SERVERPROPERTY('EngineEdition') <> 8 BEGIN /*not Azure Managed Instance*/
+	DECLARE @ErrorMessage AS nvarchar(500) = 'Telegraf - Connection string Server:'+ @@SERVERNAME + ',Database:' + DB_NAME() +' is not an Azure Managed Instance. Check the database_type parameter in the telegraf configuration.';
+	RAISERROR (@ErrorMessage,11,1)
+	RETURN
+END;
+
+DROP PROCEDURE IF EXISTS #ExecForeachDb
+/* Create a temporary procedure that iterates through all DBs that the user has access to in this managed instance, except system DBs, and execute the query specified with @queryText */
+EXEC('CREATE PROCEDURE #ExecForeachDb @queryText NVARCHAR(MAX)
+AS
+BEGIN
+	SET NOCOUNT ON;
+
+	DECLARE dbCursor CURSOR FAST_FORWARD LOCAL FOR 
+	SELECT 
+		[name] 
+	FROM sys.databases
+	WHERE HAS_DBACCESS([name]) = 1 
+		AND [name] NOT IN (''master'', ''tempdb'', ''model'', ''msdb'');
+
+	DECLARE @CurrSqlText nvarchar(max);
+	DECLARE @dbName sysname;
+
+	OPEN dbCursor;
+	FETCH NEXT FROM dbCursor INTO @dbName;
+
+	WHILE @@Fetch_Status=0 BEGIN
+		SET @CurrSqlText = N''USE ''+ QUOTENAME(@dbName ,''"'') + @queryText
+		BEGIN TRY
+			EXEC sp_executesql @CurrSqlText
+		END TRY
+		BEGIN CATCH
+			-- Check if the user has no access to DB or DB was deleted
+			IF ERROR_NUMBER() NOT IN (911, 916)
+				THROW;
+		END CATCH;
+		FETCH NEXT FROM dbCursor INTO @dbName;
+	END
+
+	CLOSE dbCursor;
+	DEALLOCATE dbCursor;
+END');
+
+/* Obtain the previous times the query was called -- they are an array because we are working with a list of DBs in the Managed Instance -- and store them in a temporary table */
+IF OBJECT_ID ('tempdb.dbo.#CachedDbCollectionTimes') IS NOT NULL DROP TABLE #CachedDbCollectionTimes;
+CREATE TABLE #CachedDbCollectionTimes (
+	[id] INT,
+	[collectionTime] DATETIMEOFFSET
+);
+
+DECLARE @XmlQueryData XML = @QueryData;
+INSERT INTO #CachedDbCollectionTimes
+SELECT
+	s.value('./id[1]', 'int') AS [id],
+	s.value('./collectionTime[1]', 'datetimeoffset') AS [collectionTime]
+FROM @XmlQueryData.nodes('/collectionTimes/ts') t(s);
+
+/* Delete records older than <Max INTERVAL_LENGTH_MINUTES> * 2 */
+DELETE FROM #CachedDbCollectionTimes WHERE collectionTime <= DATEADD(dd, -2, SYSDATETIMEOFFSET());
+
+DECLARE @SqlText NVARCHAR(MAX) = N'
+
+DECLARE @lastIntervalEndTime DATETIMEOFFSET;
+SELECT TOP 1 @lastIntervalEndTime = [collectionTime] FROM #CachedDbCollectionTimes WHERE [id] = DB_ID();
```

Comment:
Don't you need an ORDER BY here, considering you use TOP 1? There is no PK/unique constraint on the `id` column (should it be added?), so duplicates are possible. If you do add a constraint, can remove TOP 1.

Reply:
PK was added.